cli/src/main.rs

```diff
@@ -1,19 +1,219 @@
+        #[clap(long, default_value = "3001")]
+        http_port: u16,
 
-#[derive(Parser, Debug)]
+        /// The port rpc server should bind to
+        /// Default: 50051
+        /// TODO: add support to override (both here and in the development_server)
+        #[clap(long, default_value = "50051")]
+        rpc_port: u16,
+
+        /// Open the development server in your default browser after starting.
+        /// Default: true
+        /// Possible values: true, false
+        #[clap(short, long, default_value = "true")]
+        open_browser: bool,
+        // TODO: add browser override list
+    },
+    /// Stop the Mycelia development server
+    Stop {
+        /// The domain to listen on.
+        /// Default: localhost
+        /// TODO: add support to override (both here and in the development_server)
+        #[clap(short, long, default_value = "localhost")]
+        domain: String,
+
+        /// The port rpc server should bind to
+        /// Default: 50051
+        /// TODO: add support to override (both here and in the development_server)
+        #[clap(long, default_value = "50051")]
+        rpc_port: u16,
+    },
+    /// Deploy your Mycelia project
+    Deploy,
+}
+
+#[derive(Debug, Parser)]
 #[command(author, version, about, long_about = None)]
-pub struct Args {
-    /// Wasm file to invoke
-    #[arg(short, long)]
-    path: String,
+#[command(propagate_version = true)]
+struct Cli {
+    #[command(subcommand)]
+    command: Commands,
+}
+
+fn build() -> Result<(), DynError> {
+    let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+    let status = Command::new(cargo)
+        .current_dir(project_root())
+        .args(&["xtask", "build"])
+        .status()?;
+
+    if !status.success() {
+        Err(format!(
+            "`cargo xtask build` failed.
+
+Status code: {}",
+            status.code().unwrap()
+        ))?;
+    }
+
+    Ok(())
+}
+
+fn make_socket(addr: SocketAddr) -> TcpSocket {
+    let socket = TcpSocket::new_v4().unwrap();
+    socket.set_reuseaddr(true).unwrap(); // allow to reuse the addr both for connect and listen
+    socket.set_reuseport(true).unwrap(); // same for the port
+    socket.bind(addr).unwrap();
+    socket
+}
+
+async fn is_peer_connected(addr: SocketAddr) -> bool {
+    make_socket("127.0.0.1:3001".parse().unwrap())
+        .connect(dbg!(addr))
+        .await
+        .is_ok()
+}
+
+async fn start(
+    domain: &String,
+    http_port: &u16,
+    rpc_port: &u16,
+    open_browser: &bool,
+) -> Result<(), DynError> {
+    // TODO: might wanna move these prints to the development_server
+    let http_addr = format!("http://{}:{}", domain, http_port);
+    let rpc_addr = format!("http://{}:{}", domain, rpc_port);
+    println!("HTTP development server listening on {}", http_addr);
+    println!("RPC server listening on {}", rpc_addr);
+
+    let () = if !is_peer_connected("127.0.0.1:3001".parse().unwrap()).await {
```

Comment:
🤔 what if we tried just polling the `Echo` endpoint of the RPC server a few times in a loop with a delay after each poll. Its very hacky - but it'll give us confirmation that the server is started.